File: 2015/20/javascript/main.mjs
import fs from "node:fs";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";

// --- Functions ---
function divisorsSum(n) {
	let sum = 1;
	for (let i = 2; i < n; i++) {
		i * i > n ? i = n : 0;
		let b = 0;
		while (n % i < 1) {
			let c = sum * i;
			sum += c - b;
			b = c;
			n /= i;
		}
	}
	return sum;
}

function divisors(n) {
	const a = n > 1 ? [1, n] : [1];
	for (let i = 2; i <= Math.sqrt(n); i++) {
		if (n % i === 0) {
			a.push(i);
			if (i ** 2 !== n) a.push(n / i);
		}
	}
	return a;
}

// --- Parse the input file ---
const input = Number(fs.readFileSync(INPUT_FILE, 'utf8'));

// --- Puzzle 1 ---
let house = 0;
let presents = 0;
while (presents < input) {
	house++;
	presents = 10 * divisorsSum(house);
}
console.log("1. House number: ", house.toLocaleString());

// --- Puzzle 2 ---
house = 0;
presents = 0;
while (presents < input) {
	house++;
	presents = 11 * divisors(house).filter(x => x * 50 >= house).reduce((acc, x) => acc + x);
}
console.log("2. House number: ", house.toLocaleString());

File: 2015/20/python/main.py
import collections
import itertools

# --- Constants --
INPUT_FILE = "../input/input.txt"

# --- Functions ---
# Divisors unctions based on:
#   https://alexwlchan.net/2019/finding-divisors-with-python/
def prime_factors(n):
    i = 2
    while i * i <= n:
        if n % i == 0:
            n /= i
            yield i
        else:
            i += 1
    if n > 1:
        yield n

def prod(iterable):
    result = 1
    for i in iterable:
        result *= i
    return result

def get_divisors(n):
    pf = prime_factors(n)
    pf_with_multiplicity = collections.Counter(pf)
    powers = [
        [factor ** i for i in range(count + 1)]
        for factor, count in pf_with_multiplicity.items()
    ]
    for prime_power_combo in itertools.product(*powers):
        yield prod(prime_power_combo)

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
text = file.read().strip()
_input = int(text)
file.close()

# --- Puzzle 1 ---
house = 0
presents = 0
while presents < _input:
    house += 1
    presents = 10 * sum(get_divisors(house))
print(f"1. House number: {house:,d}")

# --- Puzzle 2 ---
house = 0
presents = 0
while presents < _input:
    house += 1
    presents = 11 * sum(filter(lambda x: x * 50 > house, get_divisors(house)))
print(f"2. House number: {house:,d}")
